### DISCUSSION TITLE
When you call an API, do you ever call it twice?

### DISCUSSION THREAD

**Discussion by hongmono** [rrousselGit/riverpod#3192] [Q&A] [answered]
In my case, when I call the API, it is often called twice with a difference of about 50ms.

It's hard to tell if it's a problem with riverpod or something else, I was wondering if anyone has had similar cases with riverpod. If so, is there a workaround?

**Comment by AhmedLSayed9** [comment]
No.
Without a sample code, we can't help.

**Reply by hongmono** [reply]
I don't have any example code, and it's not something I've encountered all the time, it's intermittent, so it's hard to tell. I was just wondering if you've had a similar case on Riverpod, thanks for the response.

**Reply by AhmedLSayed9** [reply]
This is critical and I've never faced such issue.

**Reply by zamiramir** [reply]
Without example code it is hard to help.

I had this issue once, when I used watch instead of read on an other provider.
This triggered a reload of the screen, which causes the FutureProvider to be called multiple times.

**Comment by mark8044** [comment, marked as answer]
I can tell you for me its happened all the time. 

In my case it was ALWAYS because of rebuilds.  Either a FutureBuilder would rebuild, or a riverpod `watch` would cause a rebuild, or I would sneeze and I would get a rebuild.

I managed to solve them all by paying attention to minimizing rebuilds.

Using riverpod `.select` was a god-send.  Look into it, it can save you not of headaches.   

When I couldn't solve the cause of the rebuild, then I resort to creating a variable for example `bool didIcalltheAPIyet = false` and then once the API is called, I would switch that to `true` and block any further calls that way.  

```
bool didIcalltheAPIyet = false;

if (didIcalltheAPIyet == false)
{
   didIcalltheAPIyet = true;
   calltheAPI();
}
```
A bit hacky but it works when you can't find the reason otherwise
